### PR TITLE
api: Do not truncate errors <1 MB

### PR DESCRIPTION
**Pull request by justinmk**
Continues #6237

- [x] limit error message size to 1 MB
- [x] convert all call sites to use `api_set_error` function instead of macro
- [x] test

### CODE REVIEW

File: test/functional/provider/python3_spec.lua [hunk 1]
```diff
@@ -30,6 +32,17 @@ describe('python3 commands and functions', function()
     eq({100, 0}, eval('g:set_by_python3'))
   end)
 
+  it('does not truncate error message <1 MB', function()
+    local very_long_symbol = string.rep('a', 1200)
+    feed_command(':silent! py3 print('..very_long_symbol..' b)')
+    -- Truncated error message would not contain this (last) line.
+    eq('SyntaxError: invalid syntax',eval('v:errmsg'))
```
Comment (brcolow) [inline test/functional/provider/python3_spec.lua]: Missing a space after the comma separating the two arguments to `eq` (I think?)

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -639,7 +638,9 @@ Boolean nvim_buf_is_valid(Buffer buffer)
     FUNC_API_SINCE(1)
 {
   Error stub = ERROR_INIT;
-  return find_buffer_by_handle(buffer, &stub) != NULL;
+  Boolean ret = find_buffer_by_handle(buffer, &stub) != NULL;
+  xfree(stub.msg);
```
Comment (ZyX-I) [inline src/nvim/api/buffer.c]: I would actually suggest using `api_free_error(Error)` or `api_clear_error(Error *)`, would be useful if we find that error handling needs to be optimized: note that e.g. Python is completely fine with EAFP approach which suggests that errors are common and expected.

**Comment by sander2** [comment]
It's great to see that the PR was not forgotten :).

> limit error message size to 1 MB

Something like this should do it:

```diff
diff --git a/src/nvim/api/private/helpers.c b/src/nvim/api/private/helpers.c
index 8bd3d51b..dded6533 100644
--- a/src/nvim/api/private/helpers.c
+++ b/src/nvim/api/private/helpers.c
@@ -977,10 +977,13 @@ void api_set_error(Error *err, ErrorType errType, const char *format, ...)
     va_start(args1, format);
     va_list args2;
     va_copy(args2, args1);
-    int len = vsnprintf(NULL, 0, format, args1);
+    size_t bufSize = (size_t)vsnprintf(NULL, 0, format, args1) + 1;
+    if (bufSize > 1024*1024) {
+        bufSize = 1024*1024;
+    }
     va_end(args1);
-    err->msg = xmalloc((size_t)len + 1);
-    vsprintf(err->msg, format, args2);
+    err->msg = xmalloc(bufSize);
+    vsnprintf(err->msg, bufSize, format, args2);
     va_end(args2);
 
     err->set = true;
```

> I would actually suggest using api_free_error(Error) or api_clear_error(Error *), would be useful if we find that error handling needs to be optimized: note that e.g. Python is completely fine with EAFP approach which suggests that errors are common and expected.

Yes, that might be nice. The `api_free_error` naming would be consistent with the other `api_free_*` functions used.

File: src/nvim/api/buffer.c (hunk 2 above)
Comment (justinmk) [inline src/nvim/api/buffer.c]: > note that e.g. Python is completely fine with EAFP approach which suggests that errors are common and expected.

Should we care about an extra call to vsnprintf?

File: test/functional/api/vim_spec.lua [hunk 3]
```diff
@@ -452,6 +452,13 @@ describe('api', function()
     ok(err:match('Invalid option name') ~= nil)
   end)
 
+  it('does not truncate error message <1 MB #5984', function()
+    local very_long_name = 'A'..('x'):rep(10000)..'Z'
+    local status, err = pcall(nvim, 'get_option', very_long_name)
+    eq(false, status)
+    eq(very_long_name, err:match('Ax+Z?'))
+  end)
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua]: This fails correctly before the change.

File: test/functional/provider/python3_spec.lua [hunk 4]
```diff
@@ -30,6 +32,15 @@ describe('python3 commands and functions', function()
...
   end)
 
+  it('does not truncate error message <1 MB', function()
+    -- XXX: Python limits the error name to 200 chars, so this test is
+    -- mostly bogus.
+    local very_long_symbol = string.rep('a', 1200)
+    feed_command(':silent! py3 print('..very_long_symbol..' b)')
+    -- Truncated error message would not contain this (last) line.
+    eq('SyntaxError: invalid syntax', eval('v:errmsg'))
+  end)
```
Comment (justinmk) [inline test/functional/provider/python3_spec.lua]: This isn't really a valid test but I left it anyway, as a sanity check.

**Pull request closed without merging**